libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -23,23 +24,147 @@
+}
+
+
 template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
-class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>
+class __fair_semaphore
 {
-  static_assert(__least_max_value <= _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>::max(), "");
+private:
+  const ptrdiff_t max_;
+  cuda::atomic<ptrdiff_t, _Sco> current;
+  cuda::atomic<ptrdiff_t, _Sco> tickets;
 
 public:
-  _LIBCUDACXX_INLINE_VISIBILITY constexpr counting_semaphore(ptrdiff_t __count = 0)
-      : _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>(__count)
-  {}
-  ~counting_semaphore() = default;
+    static_assert(__least_max_value >= 0, "The least maximum value must be a positive number");
+
+    _LIBCUDACXX_INLINE_VISIBILITY bool try_acquire()
+  {
+    ptrdiff_t available_tickets = max_ - tickets.load();
+    return (available_tickets > 0) ? true : false;
+  }
+  
+  _LIBCUDACXX_INLINE_VISIBILITY bool our_turn(ptrdiff_t ticket_num)
+  {
+    ptrdiff_t serving_ticket = current.load();
+    return ticket_num == serving_ticket;
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr ptrdiff_t max() noexcept
+  {
+    return _CUDA_VSTD::numeric_limits<ptrdiff_t>::max();
+  }
+
+  // should block until it's their turn. Right? 
+  _LIBCUDACXX_INLINE_VISIBILITY void acquire()
+  { 
+    // if there aren't any tickets available we block
+    while(!try_acquire()) {
+      simple_backoff();
+    }
+    ptrdiff_t t = tickets.fetch_add(1, cuda::memory_order_relaxed);
+    // now we're in line
+    // give the newly added value
+    while (!our_turn(t+1)) 
+    {
+      simple_backoff();
+    }
+    // blocking has concluded at this point
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY bool __acquire_slow_timed(
+    _CUDA_VSTD::chrono::nanoseconds const& __rel_time) {
+
+    return __libcpp_thread_poll_with_backoff(
+      [this]() {
+        ptrdiff_t const remaining = max_ - tickets.load(memory_order_acquire);
+        bool success = false;
+        // potential to refine this so that time between loading
+        // tickets and time between return doesn't allow possibility of
+        // another thread to alter remaining?
+        while(remaining > 0 && !success) {
+          success = tickets.compare_exchange_weak(remaining, 
+            remaining + 1,
+            memory_order_acquire,
+            memory_order_relaxed);
+        }
+        // keep looping until we finally succeed. Now that it's been added
+        // we need to make sure it's our turn
+        while(!our_turn(remaining+1)) {
+          simple_backoff();
+        }
+        return true;
+
+        // I don't think we should do the below code because it'll terminate 
+        // potentially early 
+        // return our_turn(remaining+1);
+      },
+      __rel_time);
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY void release(ptrdiff_t update = 1)
+  {
+    tickets.fetch_sub(update, memory_order_release);
```

Comment:
Why is this update `release`?

Reply:
Logically it makes more sense to be after the `current` update but curious to know if there would be some undefined behavior as it currently is? (I agree it should be changed but I'm really just curious)